### ISSUE TITLE
Native PHP extension / driver which uses fuerte and velocypack

### ISSUE THREAD

**Issue by sandrokeil** [arangodb/arangodb-php#232]
Are there any plans to create a native PHP >= 7.1 extension of this library to increase overall performance in PHP? Something like [node-arangodb-cxx](https://github.com/arangodb/node-arangodb-cxx).

Have you any idea which performance gain can be expected with this or do you think it's not worth the effort?

**Comment by jsteemann** [comment]
@sandrokeil: this is a good idea and we also thought about that a bit. 
From my experience with the native driver for node.js I would expect that the performance gains from using a native extension for request/response processing and the JSON/VelocyPack handling should be there, though they potentially would vary a great deal from use case to use case.
I'd expect to see nice speedups for use cases that process bigger requests/responses.
However, it will be a major effort to write a native extension. I do not have a good idea how long this would take as I did not write any serious PHP extension for about 10 years, and have no idea how the extension API now looks like. But it's still a huge effort to get the type mapping right and up to speed. 
If anyone volunteers to provide the basic stubs and a build environment for a native extension, that would be a good starting point, and I think we could help out then. But I do not see ourselves writing a native PHP extension in the next few weeks due to lack of time. This may change however and it may get back on our radar.

**Comment by sandrokeil** [comment]
Maybe it's a bit easier to use the [php-cpp](http://www.php-cpp.com/) project for this. They have also [async](https://github.com/CopernicaMarketingSoftware/REACT-CPP) stuff which can be used and some examples can be found there.

**Comment by sandrokeil** [comment]
There exists a project [lucassouzavieira/arangodb-php-driver](https://github.com/lucassouzavieira/arangodb-php-driver) which uses [zephir](https://zephir-lang.com/). 

@lucassouzavieira have you compared php-cpp and zephir? If so, can you say the advantages over php-cpp? Have you think about to use [fuerte](https://github.com/arangodb/fuerte) as foundation?

**Comment by lucassouzavieira** [comment]
@sandrokeil, this is exactly the idea of ​​the project with Zephir. The extension will work natively with PHP 7+. 

About [PHP-CPP](http://www.php-cpp.com/): I considered using it for the native driver, but Zephir seems to me a more interesting option. Anyway, the Zephir documentation is very broad and the integration with the Zend Engine API is very good. 

It is very easy to develop with Zephir, especially for those who already develop with PHP. It is also natively memory-safe and cross-platform.With PHP-CPP I might have to worry about interoperability between different platforms. If I had chosen PHP-CPP, I would certainly use the [fuerte](https://github.com/arangodb/fuerte) base.

**Comment by sandrokeil** [comment]
@lucassouzavieira Thanks for your reply.

>  If I had chosen PHP-CPP, I would certainly use the fuerte base.

Does this mean, that fuerte would not work with zephir? Because the [docs says](https://docs.zephir-lang.com/en/latest/config.html#extra-libs), you can also use any cpp library?

I think, if someone works on a native PHP driver, they should rely on fuerte, because of velocypack and other features. Like [node-arangodb-cxx](https://github.com/arangodb/node-arangodb-cxx) and build an abstraction like the ArangoDB PHP lib on top of that. Any idea how we can join the work to get things done?

I have created some [ArangoDB types](https://github.com/sandrokeil/arangodb-event-store/tree/master/src/Type) which are pretty handy, I guess. Not sure if it is useful for the official lib too? /cc @jsteemann 

@frankmayer You have some ArangoDB libs like [ArangoDB-PHP-Core-Guzzle](https://github.com/frankmayer/ArangoDB-PHP-Core-Guzzle). You also interested in more performance and async stuff?

**Comment by lucassouzavieira** [comment]
@sandrokeil, I agree with you regarding that new native driver support the velocypack. 

> Because the docs says, you can also use any cpp library?

We can include C libraries in the build process. If we have a C binding for the [velocypack](https://github.com/arangodb/velocypack), it can be included in the driver. Although the velocypack is written in C++, i believe it is possible to use its functions in C. [Check this](https://isocpp.org/wiki/faq/mixing-c-and-cpp#call-cpp). In this way, we can add velocypack support in the [arangodb-php-driver](https://github.com/lucassouzavieira/arangodb-php-driver).

In my view, it's better than rewriting a new native driver again.

**Comment by sandrokeil** [comment]
@lucassouzavieira / @jsteemann I've build a [php-cpp Docker image](https://hub.docker.com/r/sandrokeil/phpcpp/) to build the extension and I've created a [arangodb-php-driver respository](https://github.com/sandrokeil/arangodb-php-driver) and it looks like the extension can be build with fuerte and velocypack. :tada: 

Next step could be a velocypack and fuerte call to see if this works. Maybe we can simply port [node-arangodb-cxx](https://github.com/arangodb/node-arangodb-cxx) to PHP.

**Comment by lucassouzavieira** [comment]
@sandrokeil, I'll take a look at your repository. I believe that the Velocypack and other features of the fuerte can be included without problems in the repository that I have been working on. Only C binding would be required for the fuerte and velocypack projects in C++.

@jsteemann, can you say something about it? What do you think about that ?

**Comment by sandrokeil** [comment]
@lucassouzavieira It would be great if this works, so I guess we can go this way. I don't know if it does matters which tool we use to build the extension. Do you know if Alpine Linux is supported?

**Comment by sandrokeil** [comment]
Simple [velocypack call](https://github.com/sandrokeil/arangodb-php-driver/commit/b79bcc8d7f64d208f368c3d33897b1a510b1fd93#diff-7ec3c68a81efff79b6ca22ac1f1eabbaR46) works. [Small progress](https://github.com/sandrokeil/arangodb-php-driver/issues?utf8=%E2%9C%93&q=) but  I guess my build is ugly and I have new linking issues.

**Comment by sandrokeil** [comment]
Kudos to @martin-schilling :-) Now velocypack and fuerte are working. A [collection was created](https://github.com/sandrokeil/arangodb-php-driver/blob/master/src/main.cpp#L65-L75), but the script is [hanging](https://github.com/sandrokeil/arangodb-php-driver/issues/10). I don't know if it's a fuerte or PHP extension problem. /cc @jsteemann

**Comment by sandrokeil** [comment]
It looks like it is also possible to use the ArangoDB go driver with [php-go](https://github.com/kitech/php-go) or even with [go-php](https://github.com/deuill/go-php).

**Comment by sandrokeil** [comment]
It looks like the go variant is not an option, but [this PR](https://github.com/sandrokeil/arangodb-php-driver/pull/14/files#diff-7ec3c68a81efff79b6ca22ac1f1eabbaR66) is a game changer. :tada:
